src/Microsoft.OData.Client/DataServiceContext.cs

```diff
@@ -2483,6 +2483,61 @@ public virtual void SetLink(object source, string sourceProperty, object target)
             }
         }
 
+        /// <summary>Adds the specified link to the set of objects the <see cref="Microsoft.OData.Client.DataServiceContext" /> is tracking.</summary>
+        /// <param name="source">The source object for the new link.</param>
+        /// <param name="sourceProperty">The name of the navigation property on the source object that returns the related object.</param>
+        /// <param name="target">he object related to the source object by the new link.</param>
+        /// <exception cref="System.ArgumentNullException">When <paramref name="source" />, <paramref name="sourceProperty" /> or <paramref name="target" /> are null.</exception>
+        /// <exception cref="System.InvalidOperationException">When the specified link already exists.-or-When the objects supplied as <paramref name="source" /> or <paramref name="target" /> are in the <see cref="Microsoft.OData.Client.EntityStates.Detached" /> or <see cref="Microsoft.OData.Client.EntityStates.Deleted" /> state.-or-When <paramref name="sourceProperty" /> is not a navigation property that defines a reference to a single related object.</exception>
+        /// <remarks>
+        /// The sourceProperty MUST be a single-value navigation property.
+        /// </remarks>
+        public virtual void SetRelatedObjectLink(object source, string sourceProperty, object target)

```

Comment:
Another question about this method, it does fewer validation checks than the `EnsureRelatable` method used by `AddLink`. Is there a reason we don't do the same set of checks? Would it make sense to re-use `EnsureRelatable` and maybe pass a flag to determine whether the check is for a single or collection navigation property?

Reply:
1) The target property is the difference. I created a new method since I did not want to change how the AddLink method works.
2) We don't need to call `EnsureRelatable` since we are doing the checks inside the method.

Reply:
I have updated the summary

Reply:
Can this method be used with the normal `SaveChanges` and what will be the implications?

Reply:
I do not understand why we need to do this a single-value navigation property only and not a collection-valued navigation property?

Reply:
1) SetLink creates a `PUT` request that's why the state is `Modified`. I didn't want to change the `SetLink` logic.
2) `AddRelatedObject` works for collection navigation property. So this new method works for single-value navigation property